File: chatgpt/2022/06/main.py
# Read the input string
with open('input.txt') as f:
    s = f.read().strip()

# Keep track of the last four characters
last_four = []

# Iterate over the characters in the string
for i, c in enumerate(s):
    # Append the current character to the last_four list
    last_four.append(c)

    # If the length of the set of the last four characters is four,
    # then we have found the first start-of-packet marker
    if len(set(last_four)) == 4:
        # Print the result
        print(i + 1)
        break

    # If the length of the last_four list is more than four, remove
    # the first element from the list
    if len(last_four) >= 4:
        last_four.pop(0)

# Keep track of the last 14 characters
last_fourteen = []

# Iterate over the characters in the string
for i, c in enumerate(s):
    # Append the current character to the last_fourteen list
    last_fourteen.append(c)

    # If the length of the set of the last 14 characters is 14,
    # then we have found the first start-of-message marker
    if len(set(last_fourteen)) == 14:
        # Print the result
        print(i + 1)
        break

    # If the length of the last_fourteen list is more than 14, remove
    # the first element from the list
    if len(last_fourteen) >= 14:
        last_fourteen.pop(0)

File: chatgpt/2022/06/main.rs
use std::env;
use std::fs;

fn main() {
    // Read the first command line argument
    let part = env::args().nth(1).unwrap_or_else(|| {
        eprintln!("Error: missing part argument");
        std::process::exit(1);
    });

    // Parse the part argument
    let part = part.parse::<usize>().unwrap_or_else(|_| {
        eprintln!("Error: invalid part argument");
        std::process::exit(1);
    });

    // Read the input string
    let input = fs::read_to_string("input.txt").unwrap();
    let s = input.trim();

    // Keep track of the last 4 or 14 characters
    let mut last_four_or_fourteen = Vec::new();

    // Iterate over the characters in the string
    for (i, c) in s.chars().enumerate() {
        // Append the current character to the last_four_or_fourteen vector
        last_four_or_fourteen.push(c);

        // If the length of the set of the last 4 or 14 characters is 4 or 14,
        // then we have found the first start-of-packet or start-of-message marker
        if last_four_or_fourteen.iter().collect::<std::collections::HashSet<_>>().len() == part {
            // Print the result
            println!("{}", i + 1);
            break;
        }

        // If the length of the last_four_or_fourteen vector is more than 4 or 14, remove
        // the first element from the vector
        if last_four_or_fourteen.len() >= part {
            last_four_or_fourteen.remove(0);
        }
    }
}
